Fix VM calling convention detection for some pinvokes

AndyAyersMS opened the pull request:
The previous calling convention logic was written
assuming the `CallConv*` collection of types were
in FQTN, including an assembly. However, if the
type is being used in the assembly it is defined,
SPCL, then we fail to match the name of the type.

Fixes #62696.

AndyAyersMS committed:
Fix VM calling convention detection for some pinvokes

Change #61537 inadvertently left off a `- 1` in some string length
computations for the pinvoke calling convention tables; this caused
the x86 jit to think that some pinvokes were `stdcall` when they were
in fact `cdecl`, and that lead to GC stress failures when the stack
became unbalanced.

Fixes #62696.

AndyAyersMS commented:
Note stacks in x86 can get out of sync like this for long stretches in EBP frames without causing obvious problems.

AndyAyersMS commented:
PTAL @kunalspathak @AaronRobinsonMSFT

am11 commented on src/coreclr/vm/callconvbuilder.cpp:
```diff
@@ -60,7 +60,7 @@ namespace
 #define BASE_CALL_CONV(name, flag) { \
         MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name), \
-        STRING_LENGTH(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name)), \
+        STRING_LENGTH(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name)) - 1, \
```
nit: we can use `ARRAY_SIZE(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name))` (without -1) for these, since `STRING_LENGTH` is defined as: https://github.com/dotnet/runtime/blob/69679789fba3c377b1d14689ebab2e30a9f9ad8a/src/native/minipal/utils.h#L10

am11 reviewed (commented):
Thanks for fixing these. Sorry for the oversight.

am11 commented on src/coreclr/vm/callconvbuilder.cpp:
```diff
@@ -60,7 +60,7 @@ namespace
 #define BASE_CALL_CONV(name, flag) { \
         MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name), \
-        STRING_LENGTH(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name)), \
+        STRING_LENGTH(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name)) - 1, \
```
Should it be instead:
```suggestion
        ARRAY_SIZE(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name)) - 1, \
```
(or -2 with `STRING_LENGTH`). It is defined here:
https://github.com/dotnet/runtime/blob/0df121458740db1d0dd6cbe2c24bd0c94fb0af5e/src/native/minipal/utils.h#L7-L10
Previously, it was `lengthof(MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(name)) - 1, \` where `lengthof` was defined as `#define lengthof(rg)    (sizeof(rg)/sizeof(rg[0]))`.

AaronRobinsonMSFT replied:
I'm missing something here. As noted above the `lengthof` approach was previously identical to `ARRAY_SIZE`. The update in the referenced PR was then converted to `STRING_LENGTH`, which is `ARRAY_SIZE - 1`. The intent here, if I recall, is to get the string character count, minus the `null`. That seems to be `STRING_LENGTH`? Subtracting another 1, means we are not removing characters from the string. I've no doubt there could be an issue here, but I am missing how this change fixes a problem.

am11 replied:
You are correct, Disregard my comment. :)

`lengthof - 1` to `STRING_LENGTH` change was correct going by the definition of `lengthof` which was deleted in https://github.com/dotnet/runtime/pull/61537.

BruceForstall commented:
How could we have gone so long without seeing a failure except for in GCStress? Do we need some kind of pre-/post-call SP checking stress mode?

Would it be possible/easy to add a unit test?

jkotas commented:
> Change https://github.com/dotnet/runtime/pull/61537 inadvertently left off a - 1 in some string length
computation

I do not see the problem in this change. I have checked multiple times and lengths are same before and after the change as far as I can tell. Am I missing something?

Maybe we are reading uninitialized data somewhere and this change made the uninitialized data different?

AndyAyersMS commented:
In particular we had:

https://github.com/dotnet/runtime/blob/fee2bc2762f7eff616d274c57331299bd552f1f3/src/coreclr/vm/callconvbuilder.cpp#L20-L29

![image](https://user-images.githubusercontent.com/10121823/170302301-cd63e323-e5d7-466f-b895-812dc8d25590.png)

AaronRobinsonMSFT commented:
I think I see the issue here. The proposed fix isn't the correct solution, but it is in the vein. @AndyAyersMS The area you pointed out is the problem space for sure. Let me see what I can root out. It would seem we made an assumptions on the definition of FullyQualifiedTypeName that aren't be held up in this scenario.

AaronRobinsonMSFT commented:
Sigh... The issue here is we defined FullyQualifiedTypeName (FQTN) as described at https://docs.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names. This works in all public scenarios for the calling convention types because the FQTN defined SPCL and thus has the expected trialing `,`. However, if the type is in the assembly in question it doesn't have the trailing `,` and thus fails.

I need to think about this but it is clearly a case where we failed to consider internal use cases of the attributes. This is an interop bug.

/cc @dotnet/interop-contrib

AndyAyersMS commented:
@AaronRobinsonMSFT happy to let you take this over. And I'll update https://github.com/dotnet/runtime/issues/62696.

AaronRobinsonMSFT committed:
Revert "Fix VM calling convention detection for some pinvokes"

This reverts commit 7bacbef62551faaf079e80219058381577bd7b5d.

AaronRobinsonMSFT committed:
The previous calling convention logic was written
assuming the CallConv* collection of types were
in FQTN, including an assembly. However, if the
type is being used in the assembly it is defined,
SPCL, then we fail to match the name of the type.

jkotas commented on src/coreclr/vm/callconvbuilder.cpp:
```diff
@@ -11,9 +11,12 @@
+
+// We can include the ',' delimiter, when a type name is in Fully Qualified
+// form outside its declaring assembly - most common scenario.
+#define MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(callConvTypeName) MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME(callConvTypeName) ","
```
Instead of duplicating the type names once without `,` and second time with `,`, would it better to introduce `EqualsSimpleTypeName` matching method that succeeds when we have end-of-string or comma after the match?

AaronRobinsonMSFT replied:
I was curious if someone was going to balk at the duplicate string arrangement. I can reduce it for sure.

AaronRobinsonMSFT committed:
Review feedback.

AaronRobinsonMSFT commented:
All relevant tests have passed. The failures are from the AzDO failures and are in the outer-loop.
